engine/src/main/java/org/entando/entando/aps/system/services/component/ComponentService.java

```diff
@@ -82,5 +80,67 @@ private List<Map<String, Object>> extractReferences(ComponentUsageDetails usage,
+        listRequest.setPageSize(0); // get all elements - no pagination
+        return service.getComponentUsageDetails(componentCode, listRequest);
+    }
+    
+    @Override
+    public ComponentDeleteResponse deleteInternalComponents(List<Map<String, String>> components) {
+        ComponentDeleteResponse response = new ComponentDeleteResponse();
+        log.debug("request deletion of components '{}'", components);
+        try {
+            response.setStatus(ComponentDeleteResponse.STATUS_SUCCESS);
+            components.stream().forEach(m -> {
+                String type = m.get(ComponentValidator.TYPE_FIELD);
+                String code = m.get(ComponentValidator.CODE_FIELD);
+                log.debug("Type '{}', Object code '{}'", type, code);
+                services.stream()
+                        .filter(s -> type.equalsIgnoreCase(s.getObjectType())).findFirst()
+                        .ifPresent(service -> {
+                            Map<String, Object> properties = new HashMap<>();
+                            properties.put(REFERENCE_TYPE_PROPERTY, type);
+                            properties.put(REFERENCE_CODE_PROPERTY, code);
+                            try {
+                                service.getComponentDto(code).ifPresentOrElse(dto -> {
+                                    PagedMetadata<ComponentUsageEntity> result = this.extractUsageDetails(code, service);
+                                    if (this.checkReferences(components, result.getBody())) {
+                                        service.deleteComponent(code);
+                                        properties.put(STATUS_PROPERTY, ComponentDeleteResponse.STATUS_SUCCESS);
+                                    } else {
+                                        response.setStatus(ComponentDeleteResponse.STATUS_PARTIAL_SUCCESS);
+                                        properties.put(STATUS_PROPERTY, ComponentDeleteResponse.STATUS_FAILURE);
+                                    }
+                                }, () -> {
+                                    response.setStatus(ComponentDeleteResponse.STATUS_PARTIAL_SUCCESS);
+                                    properties.put(STATUS_PROPERTY, ComponentDeleteResponse.STATUS_FAILURE);
+                                });
+                            } catch (EntException e) {
+                                throw new RestServerError("Error extracting Component details", e);
+                            }
+                            response.getComponents().add(properties);
+                        });
+            });
+        } catch (Exception e) {
+            throw new RestServerError("Error deleting components", e);
+        }
+        return response;
+    }
+    
+    private boolean checkReferences(List<Map<String, String>> componentsToDelete, List<ComponentUsageEntity> extractedReferences) {
+        for (ComponentUsageEntity ref : extractedReferences) {
```

Comment:
computal complexity too hig and this code is a mix from imperative and functional, use somting like
`        Map<String,String> refs = extractedReferences.stream().collect(Collectors.toMap(cu -> cu.getType+cu.getCode(), cu -> cu.getType+cu.getCode()));
        return extractedReferences.stream()
            .anyMatch(m -> refs.containsKey(ctd.get(ComponentValidator.TYPE_FIELD)+ctd.get(ComponentValidator.CODE_FIELD)));
`